### PR TITLE
CLUS-3235 adding 'dbversions' subcommand

### PR DESCRIPTION
**Pull request by alvaro-vinuela**

### CODE REVIEW

File: doc/s9s.dbversions.1 [hunk 1]
```diff
@@ -0,0 +1,69 @@
+.TH S9S-DBVERSIONS 1 "December 19, 2023"
```
Comment (ivashchenko-volodymyr) [inline doc/s9s.dbversions.1:1]: Did you do the check of spelling in the document? If not yet please do it. Thank you.

File: doc/s9s.dbversions.1 [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+The list operations: 'getDbVersions' (--list-versions) and 'getDb3dVersions' (--list-3d-versions)
+can be used to get the list of available versions and they need following Arguments
+
+.TP
+.BI \-\^\-cluster-type= CLUSTER_TYPE_NAME 
+This option must be used on get versions operations. 
+The possible values to use are returned by the --cluster-types operation (on upper or lower case).
+
+.TP
+.BI \-\^\-provider= PROVIDER_NAME
```
Comment (ivashchenko-volodymyr) [inline doc/s9s.dbversions.1:67]: Please check the parameter name.

File: libs9s/s9srpcreply.cpp [hunk 3]
```diff
@@ -347,6 +347,46 @@ S9sRpcReply::users()
...
 }
 
+S9sVariantList
+S9sRpcReply::dbVersions()
+{
+    S9sVariantList  theList;
+    
+    if (contains("db_versions"))
+        theList = operator[]("db_versions").toVariantList();
+    else if (contains("db_versions"))
```
Comment (ivashchenko-volodymyr) [inline libs9s/s9srpcreply.cpp:357]: same `contains("db_versions")` condition in `if` and its `esle if`, why it is so?

File: libs9s/s9srpcreply.cpp [hunk 4]
```diff
@@ -347,6 +347,46 @@ S9sRpcReply::users()
...
+}
+
+S9sVariantList
+S9sRpcReply::clusterTypes()
+{
+    S9sVariantList  theList;
+    
+    if (contains("cluster_types"))
+        theList = operator[]("cluster_types").toVariantList();
+    else if (contains("cluster_types"))
```
Comment (ivashchenko-volodymyr) [inline libs9s/s9srpcreply.cpp:370]: same `contains("cluster_types")` condition in `if` and its `esle if`, why it is so?

File: libs9s/s9srpcreply.cpp [hunk 5]
```diff
@@ -347,6 +347,46 @@ S9sRpcReply::users()
...
+}
+
+S9sVariantList
+S9sRpcReply::vendors()
+{
+    S9sVariantList  theList;
+    
+    if (contains("vendors"))
+        theList = operator[]("vendors").toVariantList();
+    else if (contains("vendors"))
```
Comment (ivashchenko-volodymyr) [inline libs9s/s9srpcreply.cpp:383]: same contains("vendors") condition in if and its esle if, why it is so?

**Review by ivashchenko-volodymyr — CHANGES_REQUESTED** [review]
Hi Alvaro, I have added some comments. Please take a look. I will continue my review after your answers. Thanks.

File: libs9s/s9srpcreply.cpp (hunk 5 above)
Comment (alvaro-vinuela) [reply libs9s/s9srpcreply.cpp:383]: you are actually right, I copied it from S9sRpcReply::clusters() I will change that

**Comment by alvaro-vinuela** [comment]
@ivashchenko-volodymyr I think it is ready to continue review. Thanks for your comments.

File: doc/s9s.dbversions.1 [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+cluster type and vendor.
+
+.TP
+.B --list-3d-versions
+Lists the available 3 digits versions supported by ClusterControl for a given 
+cluster type and vendor (if supported). If not, a bad request will be returned.
+
+.B EXAMPLE
+.nf
+s9s dbversions --list-versions --cluster-type galera --vendor mariadb
```
Comment (ivashchenko-volodymyr) [inline doc/s9s.dbversions.1:76]: Now I get `Invalid operation 'getDbVersions' on '/v2/info'.`:
```
19:42:41 s9s-tools$ ./s9s/s9s dbversions --list-versions --cluster-type galera --vendor mariadb
Invalid operation 'getDbVersions' on '/v2/info'.
19:42:50 s9s-tools$ gitstbr
On branch CLUS-3235
Your branch is up to date with 'origin/CLUS-3235'.

nothing to commit, working tree clean
---------------

* 05dbc053 - (3 hours ago) CLUS-3235 remove unneded iostream - alvaro-vinuela (HEAD -> CLUS-3235, origin/CLUS-3235)
* 14b92b67 - (3 hours ago) CLUS-3235 speeling check on man page and other PR comments - alvaro-vinuela
* 6e477a60 - (28 hours ago) CLUS-3235 adding cmon-user and pass options to 'dbversions' subcommand - alvaro-vinuela
* 3296852e - (29 hours ago) CLUS-3235 adding 'dbversions' subcommand - alvaro-vinuela
```
Did you test the command? Does it work for you?

**Review by ivashchenko-volodymyr — CHANGES_REQUESTED** [review]

File: doc/s9s.dbversions.1 (hunk 6 above)
Comment (ivashchenko-volodymyr) [reply doc/s9s.dbversions.1:76]: I probably need `cmon` with the related changes to get the command working

Comment (ivashchenko-volodymyr) [reply doc/s9s.dbversions.1:76]: Works:
```
23:21:31 ~$ s9s dbversions --list-versions --cluster-type galera --vendor mariadb
10.4
10.5
10.6
10.7
10.8
10.9
10.10
10.11
23:22:00 ~$ 
```
:)

Comment (ivashchenko-volodymyr) [reply doc/s9s.dbversions.1:76]: Will continue review tomorrow...

Comment (ivashchenko-volodymyr) [reply doc/s9s.dbversions.1:76]: FYI:
```
23:23:48 ~$ s9s dbversions --list-3d-versions --cluster-type galera --vendor mariadb
10.4.32
10.5.23
10.6.16
10.10.7
10.11.6
11.0.4
11.1.3
11.2.2
11.3.1
23:24:00 ~$ s9s dbversions --list-3d-versions --cluster-type mongodb --vendor 10gen
Invalid operation 'getDb3dVersions' on '/v2/dbversions'.
23:24:14 ~$ s9s dbversions --list-versions --cluster-type mongodb --vendor 10gen
4.4
5.0
6.0
23:24:24 ~$ 
```

**Comment by ivashchenko-volodymyr** [comment]
Is it possible to have better message for cases like:
```
14:45:47 s9s-tools$ s9s dbversions --list-3d-versions --cluster-type mongodb --vendor 10gen
Invalid operation 'getDb3dVersions' on '/v2/dbversions'.
```
?
Some message with `operation is not supported for the cluster type`...

File: libs9s/s9soptions.cpp [hunk 7]
```diff
@@ -10280,6 +10378,273 @@ S9sOptions::checkOptionsDbSchema()
...
+ */
+bool
+S9sOptions::readOptionsDbVersions(
+        int argc,
+        char *argv[])
+{
+    int c;
+    struct option long_options[] =
+            {
+                    // Generic Options
```
[the remark this answers is not in the document]
Comment (alvaro-vinuela) [reply libs9s/s9soptions.cpp:10392]: Yes. Removing config-file which is useless in this case

**Review by ivashchenko-volodymyr — APPROVED** [review]

**Review by ivashchenko-volodymyr — APPROVED** [review]

**Pull request merged**